File: valid-anagram/Solution242.java
// https://leetcode.com/problems/valid-anagram/

class Solution242 {

    // T: O(n), S: O(1)
    public boolean isAnagram(String s, String t) {
        if (s.length() != t.length()) {
            return false;
        }
        var counts = new int[128];
        for (int i = 0; i < s.length(); i++) {
            counts[s.charAt(i)]++;
            counts[t.charAt(i)]--;
        }
        for (int count : counts) {
            if (count != 0) {
                return false;
            }
        }
        return true;
    }

    public static void main(String[] args) {
        var sol = new Solution242();
        var s1 = "anagram";
        var t1 = "nagaram";
        System.out.println(sol.isAnagram(s1, t1)); // true
	    var s2 = "rat";
        var t2 = "car";
        System.out.println(sol.isAnagram(s2, t2)); // false
    }

}

File: valid-anagram/solution242.py
# https://leetcode.com/problems/valid-anagram/

from collections import Counter


class Solution242:

    # T: O(n), S: O(k)
    def isAnagram(self, s: str, t: str) -> bool:
        return Counter(s) == Counter(t)


def main() -> None:
    sol = Solution242()
    s = "anagram"
    t = "nagaram"
    print(sol.isAnagram(s, t))  # true
    s = "rat"
    t = "car"
    print(sol.isAnagram(s, t))  # false


if __name__ == "__main__":
    main()
